client/components/site-icon/index.jsx

```diff
@@ -70,6 +75,19 @@ const SiteIcon = React.createClass( {
 	}
 } );
 
-export default connect( ( state, { siteId } ) => (
-	siteId ? { site: getSite( state, siteId ) } : {}
-) )( SiteIcon );
+export default connect( ( state, { site, siteId, imgSize } ) => {
+	// Until sites state is completely within Redux, we provide compatibility
+	// in cases where site object is passed to use the icon.img property as URL
+	if ( site ) {
+		return {
+			iconUrl: get( site, 'icon.img' )
+		};
+	}
+
+	// Otherwise, assume we want to perform the lookup in Redux state
+	// exclusively using the site ID
+	return {
+		site: getSite( state, siteId ),
+		iconUrl: getSiteIconUrl( state, siteId, imgSize )
```

Comment:
>This selector relies on two sub-state trees: sites and media, Shouldn't we add Queries to the component?

You raise a good point here. In an [earlier comment](https://github.com/Automattic/wp-calypso/pull/10016#issuecomment-267364290), I noted a worry about the number of ytihhrt from `<SiteIcon />` components since they're used for each of the user's sites in the site selector.

If we were to render a `<QuerySites />`, maybe it should be limited to the condition that the site data is not yet known (and that it's passed as a site ID, i.e. Redux)?

Regarding `<QueryMedia />`, we benefit from the fallback behavior which uses the site's `site.icon.img` property (a URL string) if the media by ID is not known. The ID only changes when updating the site icon within Calypso, at which point we've received the media explicitly (in `<SiteIconSetting />)`. I _think_ this should be enough that it would never be inaccurate, though admittedly depends on a tangled set of assumptions.